zaza/charm_tests/test_utils.py

```diff
@@ -71,23 +71,43 @@ def config_change(self, default_config, alternate_config):
         :param alternate_config: Dict of charm settings to change to
         :type alternate_config: dict
         """
-        logging.debug('Changing charm setting to {}'.format(alternate_config))
-        model.set_application_config(
-            self.application_name,
-            alternate_config,
-            model_name=self.model_name)
+        # we need to compare config values to what is already applied before
+        # attempting to set them.  otherwise the model will behave differently
+        # than we would expect while waiting for completion of the change
+        _app_config = model.get_application_config(self.application_name)
+        app_config = {}
+        for k, v in _app_config.items():
```

Comment:
Why are we looking at all configuration keys, values rather than just what is expected to change in alternate_conifg?

for key in alternate_config.keys():
   if _app_config.get(key) == alternate_config[key]:
       # Do something or don't

Reply:
There are some problems that make this not as easy as it seems.

1. The function accepts a dict of possibly multiple configuration options to set, we need to check all of them.

2. The structure returned from `zaza.model.get_application_config` looks like this:
```
{'preferred-api-version': {'description': 'Use this keystone api version for '
                                          'keystone endpoints and advertise '
                                          'this\n'
                                          'version to identity client charms.  '
                                          'For OpenStack releases < Queens '
                                          'this\n'
                                          'option defaults to 2; for Queens or '
                                          'later it defaults to 3.\n',
                           'source': 'user',
                           'type': 'int',
                           'value': 3}}
```
Before we can compare that to the simpler dict we get from our caller we need to convert it. I guess we could convert only the keys we are interested in and not all of them before the comparison.

3. Types do not work like they appear,if you attempt to set a config using a `int` as juju describes you get a error like this:
```
  File "/home/ubuntu/src/keystone/.tox/func/lib/python3.5/site-packages/zaza/model.py", line 442, in async_set_application_config
    .set_config(configuration))
  File "/home/ubuntu/src/keystone/.tox/func/lib/python3.5/site-packages/juju/application.py", line 308, in set_config
    return await app_facade.Set(self.name, config)
  File "/home/ubuntu/src/keystone/.tox/func/lib/python3.5/site-packages/juju/client/facade.py", line 412, in wrapper
    reply = await f(*args, **kwargs)
  File "/home/ubuntu/src/keystone/.tox/func/lib/python3.5/site-packages/juju/client/_client5.py", line 655, in Set
    reply = await self.rpc(msg)
  File "/home/ubuntu/src/keystone/.tox/func/lib/python3.5/site-packages/juju/client/facade.py", line 537, in rpc
    result = await self.connection.rpc(msg, encoder=TypeEncoder)
  File "/home/ubuntu/src/keystone/.tox/func/lib/python3.5/site-packages/juju/client/connection.py", line 316, in rpc
    raise errors.JujuAPIError(result)
juju.errors.JujuAPIError: json: cannot unmarshal number into Go struct field ApplicationSet.options of type string
```
Thus the conversion to `str` across the board.